webgpu.h

```diff
@@ -0,0 +1,1025 @@
+    const void* nextInChain;
+    wgpuStructureType structureType;
+    wgpuTexture texture;
+    uint32_t mipLevel;
+    uint32_t arrayLayer;
+    wgpuOrigin3D origin;
+} wgpuTextureCopyView;
+
+typedef struct wgpuTextureDescriptor {
+    const void* nextInChain;
+    wgpuStructureType structureType;
+    wgpuTextureUsageBit usage;
+    wgpuTextureDimension dimension;
+    wgpuExtent3D size;
+    uint32_t arrayLayerCount;
+    wgpuTextureFormat format;
+    uint32_t mipLevelCount;
+    uint32_t sampleCount;
+} wgpuTextureDescriptor;
+
+typedef struct wgpuVertexBufferDescriptor {
+    uint64_t stride;
+    wgpuInputStepMode stepMode;
+    uint32_t attributeCount;
+    wgpuVertexAttributeDescriptor const * attributes;
+} wgpuVertexBufferDescriptor;
+
+typedef struct wgpuRenderPassDescriptor {
+    uint32_t colorAttachmentCount;
+    const wgpuRenderPassColorAttachmentDescriptor* const * colorAttachments;
+    wgpuRenderPassDepthStencilAttachmentDescriptor const * depthStencilAttachment;
+} wgpuRenderPassDescriptor;
+
+typedef struct wgpuVertexInputDescriptor {
+    const void* nextInChain;
+    wgpuStructureType structureType;
+    wgpuIndexFormat indexFormat;
+    uint32_t bufferCount;
+    wgpuVertexBufferDescriptor const * buffers;
+} wgpuVertexInputDescriptor;
+
+typedef struct wgpuRenderPipelineDescriptor {
+    const void* nextInChain;
+    wgpuStructureType structureType;
+    wgpuPipelineLayout layout;
+    wgpuPipelineStageDescriptor const * vertexStage;
+    wgpuPipelineStageDescriptor const * fragmentStage;
+    wgpuVertexInputDescriptor const * vertexInput;
+    wgpuPrimitiveTopology primitiveTopology;
+    wgpuRasterizationStateDescriptor const * rasterizationState;
+    uint32_t sampleCount;
+    wgpuDepthStencilStateDescriptor const * depthStencilState;
+    uint32_t colorStateCount;
+    const wgpuColorStateDescriptor* const * colorStates;
+    uint32_t sampleMask;
+    bool alphaToCoverageEnabled;
+} wgpuRenderPipelineDescriptor;
+
+
+typedef void (*wgpuBufferCreateMappedCallback)(wgpuBufferMapAsyncStatus status,
+                                               wgpuCreateBufferMappedResult result,
+                                               void* userdata);
+typedef void (*wgpuBufferMapReadCallback)(wgpuBufferMapAsyncStatus status,
+                                          const void* data,
+                                          uint64_t dataLength,
+                                          void* userdata);
+typedef void (*wgpuBufferMapWriteCallback)(wgpuBufferMapAsyncStatus status,
+                                           void* data,
+                                           uint64_t dataLength,
+                                           void* userdata);
+typedef void (*wgpuFenceOnCompletionCallback)(wgpuFenceCompletionStatus status, void* userdata);
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// Procs of BindGroup
+typedef void (*wgpuProcBindGroupReference)(wgpuBindGroup bindGroup);
```

Comment:
Does this have to be a part of the API header? Looks like an implementation detail to me.

Reply:
Every WebGPU implementation will need to implement something like this to avoid freeing data in use by the GPU, and having a refcount helps composing together software that uses webgpu.h

Reply:
We don't have those procs, and we don't free data in use by GPU.

Reply:
Sorry I wasn't clear. I expect every WebGPU implementation will have internal refcounting to avoid freeing data on the GPU (and I see wgpu-rs have ref_count even though I'm not familiar enough with Rust to understand all of what's happening there). So why not expose this to the application since it helps compose things together?

This is also slightly closer to the JS semantics with GC keeping the object alive as long as there is a reference to it.

Reply:
We haven't considered exposing the refcounts to the user:
  - for JS, it already knows when to call destroy()
  - for Rust, we use ownership and borrowing instead of refcounts on the API side

Perhaps, you could do it as an extension?

Reply:
Since WebGPU objects are internally synchronized doesn't it make sense to have them implement `Clone + Sync` and this would kinda require a refcount? (I don't know Rust too well so maybe this is stupid)

Reply:
Only `Device` in WebGPU is truly internally synchronized.
In general, if something is internally synchronized, that means `Sync` is fine, but `Clone` has nothing to do with it. If a user wants to share the `Device` between multiple threads, they just wrap it in `Arc<Device>` and go from there.

Reply:
Using `Arc` forces the way refcounting is done, which doesn't match what Dawn is doing (it's C++ and the refcount is at an offset while it is at offset 0 for `Arc`) nor what `wgpu-native` because of the `HUB` structure they use.

That said the same `Arc` semantics could be done with a `Sync + Clone` object that uses `webgpu.h`'s refcounting mechanisms to implement `Clone`.